services/motion/builtin/builtin.go

```diff
@@ -286,302 +284,54 @@ func (ms *builtIn) MoveOnGlobe(
 ) (bool, error) {
 	operation.CancelOtherWithLabel(ctx, builtinOpLabel)
 
-	positionPollingPeriod := time.Duration(1000/motionCfg.PositionPollingFreqHz) * time.Millisecond
-	obstaclePollingPeriod := time.Duration(1000/motionCfg.ObstaclePollingFreqHz) * time.Millisecond
-
-	movementSensor, ok := ms.movementSensors[movementSensorName]
-	if !ok {
-		return false, resource.DependencyNotFoundError(movementSensorName)
-	}
-
-	planRequest, kb, err := ms.newMoveOnGlobeRequest(ctx, componentName, destination, movementSensor, obstacles, motionCfg, extra)
+	moveRequest, err := ms.newMoveOnGlobeRequest(ctx, componentName, destination, movementSensorName, obstacles, motionCfg, extra)
 	if err != nil {
 		return false, err
 	}
 
-	// successChan is for sending messages that relate to the base meeting the success criteria for the request
-	successChan := make(chan bool, 1)
-	defer close(successChan)
-
-	// replanChan is for sending messages that signal that we should stop motion and replan
-	// initialize with a true signal so that we immediately start planning
-	replanChan := make(chan bool, 1)
-	defer close(replanChan)
-	replanChan <- true
-
-	// errChan is for sending error messages encountered during the move
-	errChan := make(chan error)
-	defer close(errChan)
-
-	// create a waitgroup to manage goroutines and wait on their completion before exiting the function
-	var backgroundWorkers sync.WaitGroup
-	defer backgroundWorkers.Wait()
-
-	// create a cancellable context and ensure that it is cancelled before exiting
-	var cancelCtx context.Context
-	var cancelFn context.CancelFunc
-	defer func() {
-		if cancelFn != nil {
-			cancelFn()
-		}
-	}()
-
-	// helper function to manage polling functions
-	startPolling := func(ctx context.Context, period time.Duration, fn func(context.Context) error) {
-		backgroundWorkers.Add(1)
-		goutils.ManagedGo(func() {
-			ticker := time.NewTicker(period)
-			defer ticker.Stop()
-			for {
-				if err := ctx.Err(); err != nil {
-					return
-				}
-				select {
-				case <-ctx.Done():
-					return
-				case <-ticker.C:
-					if err := fn(ctx); err != nil {
-						errChan <- err
-					}
-					return
-				}
-			}
-		}, backgroundWorkers.Done)
-	}
-
 	// start the loop that (re)plans when something is read from the replan channel
 	// and exits when something is read from the success channel
 	for {
-		if err := ctx.Err(); err != nil {
+		ma := newMoveAttempt(ctx, moveRequest)
+		ma.start()
+
+		// this ensures that if the context is cancelled we always return early at the top of the loop
+		if ctx.Err() != nil {
+			ma.cancel()
 			return false, err
 		}
+
```

Comment:
1. I think this should be moved to the top of the loop
2. I don't think this is going to return the error from the context.
```suggestion
// this ensures that if the context is cancelled we always return early at the top of the loop
		if err := ctx.Err(); err != nil {
			ma.cancel()
			return false, err
		}
		ma := newMoveAttempt(ctx, moveRequest)
		ma.start()
```